nav2_ceres_costaware_smoother/include/nav2_ceres_costaware_smoother/smoother.hpp

```diff
@@ -0,0 +1,396 @@
+// Copyright (c) 2021 RoboTech Vision
+// Copyright (c) 2020, Samsung Research America
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License. Reserved.
+
+#ifndef NAV2_CERES_COSTAWARE_SMOOTHER__SMOOTHER_HPP_
+#define NAV2_CERES_COSTAWARE_SMOOTHER__SMOOTHER_HPP_
+
+#include <cmath>
+#include <vector>
+#include <iostream>
+#include <memory>
+#include <queue>
+#include <utility>
+#include <deque>
+#include <limits>
+#include <algorithm>
+
+#include "nav2_ceres_costaware_smoother/smoother_cost_function.hpp"
+#include "nav2_ceres_costaware_smoother/utils.hpp"
+
+#include "ceres/ceres.h"
+#include "Eigen/Core"
+
+namespace nav2_ceres_costaware_smoother
+{
+
+/**
+ * @class nav2_smac_planner::Smoother
+ * @brief A Conjugate Gradient 2D path smoother implementation
+ */
+class Smoother
+{
+public:
+  /**
+   * @brief A constructor for nav2_smac_planner::Smoother
+   */
+  Smoother() {}
+
+  /**
+   * @brief A destructor for nav2_smac_planner::Smoother
+   */
+  ~Smoother() {}
+
+  /**
+   * @brief Initialization of the smoother
+   * @param params OptimizerParam struct
+   */
+  void initialize(const OptimizerParams params)
+  {
+    _debug = params.debug;
+
+    _options.linear_solver_type = ceres::DENSE_QR;
```

Comment:
How does `SPARSE_NORMAL_CHOLESKY` perform? I know some settings are so bad that its not worth exposing them to users, but just checking if we should strictly prevent users from selecting another legitimate option

Reply:
Seems like it is quite usable. For now, I just ran tests with `SPARSE_NORMAL_CHOLESKY` (not trying to tune) and it seems to perform similarly in most tests, and even better (and much faster) in the cusp tests. I can add a parameter to pick between these two solver types and make `SPARSE_NORMAL_CHOLESKY` default